cmd/zoekt-sourcegraph-indexserver/index.go

```diff
@@ -156,30 +185,84 @@ func gitIndex(o *indexArgs, runCmd func(*exec.Cmd) error) error {
-	// the server to have configured both uploadpack.allowAnySHA1InWant and
-	// uploadpack.allowFilter. (See gitservice.go in the Sourcegraph repository)
-	fetchArgs := []string{"-C", gitDir, "-c", "protocol.version=2", "fetch", "--depth=1", o.CloneURL}
-	var commits []string
-	for _, b := range o.Branches {
-		commits = append(commits, b.Version)
+	defer func() {
+		success := strconv.FormatBool(allFetchesSucceeded)
+		name := repoNameForMetric(o.Name)
+		metricFetchDuration.WithLabelValues(success, name).Observe(fetchDuration.Seconds())
+	}()
+
+	var fetchCommits = func(commits []string) error {
+		// We shallow fetch each commit specified in zoekt.Branches. This requires
+		// the server to have configured both uploadpack.allowAnySHA1InWant and
+		// uploadpack.allowFilter. (See gitservice.go in the Sourcegraph repository)
+		fetchArgs := []string{"-C", gitDir, "-c", "protocol.version=2", "fetch", "--depth=1", o.CloneURL}
+		fetchArgs = append(fetchArgs, commits...)
+
+		cmd = exec.CommandContext(ctx, "git", fetchArgs...)
+		cmd.Stdin = &bytes.Buffer{}
+
+		start := time.Now()
+		err := runCmd(cmd)
+		fetchDuration += time.Since(start)
+
+		if err != nil {
+			allFetchesSucceeded = false
+			return err
+		}
+
+		successfullyFetchedCommitsCount += len(commits)
+		return nil
 	}
-	fetchArgs = append(fetchArgs, commits...)
 
-	cmd = exec.CommandContext(ctx, "git", fetchArgs...)
-	cmd.Stdin = &bytes.Buffer{}
+	var branchCommits []string
+	for _, b := range o.Branches {
+		branchCommits = append(branchCommits, b.Version)
+	}
 
-	err = runCmd(cmd)
-	fetchDuration := time.Since(fetchStart)
+	err = fetchCommits(branchCommits)
 	if err != nil {
-		metricFetchDuration.WithLabelValues("false", repoNameForMetric(o.Name)).Observe(fetchDuration.Seconds())
 		return err
 	}
 
-	metricFetchDuration.WithLabelValues("true", repoNameForMetric(o.Name)).Observe(fetchDuration.Seconds())
-	debug.Printf("fetched git data for %q (%d commit(s)) in %s", o.Name, len(commits), fetchDuration)
+	if o.UseDelta {
+		// Try fetching prior commits for delta builds
+		// If we're unable to fetch prior commits, we continue anyway
+		// knowing that zoekt-git-index will fall back to a "full" normal build
+
+		var priorCommits []string
+
+		existingRepository, found, err := findRepositoryMetadata(o)
+		if err != nil {
+			return fmt.Errorf("delta build: failed to get repository metadata: %w", err)
+		}
+
+		if found {
+			for _, b := range existingRepository.Branches {
+				priorCommits = append(priorCommits, b.Version)
+			}
+		}
+
+		if len(priorCommits) > 0 {
+			err = fetchCommits(priorCommits)
+			if err != nil {
+				repositoryName := buildOptions.RepositoryDescription.Name
+				repositoryID := buildOptions.RepositoryDescription.ID
+
+				var formattedCommits []string
+				for _, b := range existingRepository.Branches {
+					formattedCommits = append(formattedCommits, fmt.Sprintf("%s@%s", b.Name, b.Version))
+				}
```

Comment:
do we maybe need a helper somewhere to generat a human readable description of buildOptions? This isn't the only place we have quite a bit of code to generate a log message. I think even before your PRs we had something somewhere which would generate these descriptions.

Reply:
I took a stab at this in [`450ce66` (#311)](https://github.com/sourcegraph/zoekt/pull/311/commits/450ce665c215deb3e0b752e18efa6dbe76113d09). 

Most of the stringer implementations I found simply used `fmt.Sprintf("+%v")` or similar, which I think is a lot of extra noise for what we want to convey here. In general I think it's hard to create a Stringer implementation for a struct with so many fields in it. 

I made a slightly smaller change: give `zoekt.RepositoryBranch` a `String()` implementation and leverage that here along with cleaning up `fetch`'s call site a bit more. I also considered making a `type zoekt.RepositoryBranches []zoekt.RepositoryBranch` and giving it its own stringer, but I thought it seemed like a bit much to introduce a new type just for this one method. 

Let me know if you disagree and I can take another stab at refactoring this.

Reply:
LGTM. Your string method doesn't have to represent all fields, could just represent what we care about. Or you could use a method other than `String` on `BuildOptions`